Wikipedia/Code/NotificationsCenterModelController.swift

```diff
@@ -0,0 +1,80 @@
+
+import Foundation
+import WMF
+
+protocol NotificationsCenterModelControllerDelegate: AnyObject {
+    func reloadCellWithViewModelIfNeeded(viewModel: NotificationsCenterCellViewModel)
+}
+
+//Keeps track of the RemoteNotification managed objects and NotificationCenterCellViewModels that power Notification Center in a performant way
+final class NotificationsCenterModelController {
+    
+    typealias RemoteNotificationKey = String
+
+    private var cellViewModelsDict: [RemoteNotificationKey: NotificationsCenterCellViewModel] = [:]
+    private var cellViewModels: Set<NotificationsCenterCellViewModel> = []
+    
+    weak var delegate: NotificationsCenterModelControllerDelegate?
+    private let languageLinkController: MWKLanguageLinkController
+    
+    init(languageLinkController: MWKLanguageLinkController, delegate: NotificationsCenterModelControllerDelegate?) {
+        self.delegate = delegate
+        self.languageLinkController = languageLinkController
+    }
+    
+    func addNewCellViewModelsWith(notifications: [RemoteNotification], editMode: Bool) {
+        for notification in notifications {
+
+            //Instantiate new view model and insert it into tracking properties
+            
+            guard let key = notification.key,
+                  let newCellViewModel = NotificationsCenterCellViewModel(notification: notification, languageLinkController: languageLinkController, editMode: editMode) else {
+                continue
+            }
+            
+            cellViewModelsDict[key] = newCellViewModel
+            cellViewModels.insert(newCellViewModel)
+        }
+    }
+    
+    func updateCurrentCellViewModelsWith(updatedNotifications: [RemoteNotification]? = nil, editMode: Bool) {
+
+        let cellViewModelsToUpdate: [NotificationsCenterCellViewModel]
+        
+        if let updatedNotifications = updatedNotifications {
+            
+            //Find existing cell view models via tracking properties
+            cellViewModelsToUpdate = updatedNotifications.compactMap { notification in
+                
+                guard let key = notification.key else {
+                    return nil
+                }
+                
+                return cellViewModelsDict[key]
+                
+            }
+            
+        } else {
+            cellViewModelsToUpdate = Array(cellViewModels)
+        }
+        
+        cellViewModelsToUpdate.forEach {
+            $0.updateWith(editMode: editMode)
+            self.delegate?.reloadCellWithViewModelIfNeeded(viewModel: $0)
+        }
+    }
+    
+    var fetchOffset: Int {
+        return cellViewModels.count
```

Comment:
@staykids yeah this is annoying because Core Data doesn't play great with Swift, but the `key` attribute in the data model is set to non-optional, so Core Data shouldn't allow any records saved with that missing. Unfortunately the class still generates that property as an optional, but I don't think it will be possible to actually have this `key` value be nil.

I do still want to do further testing on this to make sure I'm thinking of the fetch offset right, particularly if we insert a new record at the top from a refresh, then fetch the next page. I think `DiffableDataSource` should hide any duplicate cells automatically, but I want to make sure we aren't skipping any records in the database in some instances. I'll report back if I see any bugs.

Reply:
@staykids I did some light testing on a few things to try and catch any issues with this:

1. After the first page is fetched and displayed, I faked a new refresh that caused cells to inserted at the top of the list, then I scrolled down until it fetches the next page. I compared the new records inserted at the bottom with the records in the sqlite database, and didn't see any records skipped. So I think we're good on that.
2. I changed `fetchOffset` to return `cellViewModels.count - 3` to force the database to return duplicate records that are already displaying. Because `cellViewModels` is a Set, it didn't add any duplicate records and the data source stayed unique. I didn't see any dupes shown after the snapshot updated. I think we're good here too.
3. I tried to insert a `nil` key [here](https://github.com/wikimedia/wikipedia-ios/blob/ee33257b9879699fc74499fc82625b877900d166/WMF%20Framework/Remote%20Notifications/Model/RemoteNotificationsModelController.swift#L197) into the database, but it failed and the sqlite db remained empty.

Hope this helps! Let me know if you think I missed anything here.

Reply:
Awesome – thanks for this investigative work!